runtime/file.cc

```diff
@@ -0,0 +1,322 @@
+  if (fd_ >= 0) {
+    if (::close(fd_) != 0) {
+      handler.SignalErrno();
+    }
+  }
+  path_ = SaveDefaultCharacter(path, pathLength, handler);
+  if (!path_.get()) {
+    handler.Crash(
+        "FILE= is required unless STATUS='OLD' and unit is connected");
+  }
+  fd_ = ::open(path_.get(), flags, 0600);
+  if (fd_ < 0) {
+    handler.SignalErrno();
+  }
+  pending_.reset();
+  knownSize_.reset();
+}
+
+void OpenFile::Close(
+    const char *status, std::size_t statusLength, IoErrorHandler &handler) {
+  CriticalSection criticalSection{lock_};
+  CheckOpen(handler);
+  pending_.reset();
+  knownSize_.reset();
+  static const char *statuses[]{"KEEP", "DELETE", nullptr};
+  switch (IdentifyValue(status, statusLength, statuses)) {
+  case 0: break;
+  case 1:
+    if (path_.get()) {
+      ::unlink(path_.get());
+    }
+    break;
+  default:
+    if (status) {
+      handler.Crash(
+          "Invalid STATUS='%.*s'", status, static_cast<int>(statusLength));
+    }
+  }
+  path_.reset();
+  if (fd_ >= 0) {
+    if (::close(fd_) != 0) {
+      handler.SignalErrno();
+    }
+    fd_ = -1;
+  }
+}
+
+std::size_t OpenFile::Read(Offset at, char *buffer, std::size_t minBytes,
+    std::size_t maxBytes, IoErrorHandler &handler) {
+  if (maxBytes == 0) {
+    return 0;
+  }
+  CriticalSection criticalSection{lock_};
+  CheckOpen(handler);
+  if (!Seek(at, handler)) {
+    return 0;
+  }
+  if (maxBytes < minBytes) {
+    minBytes = maxBytes;
+  }
+  std::size_t got{0};
+  while (got < minBytes) {
+    auto chunk{::read(fd_, buffer + got, maxBytes - got)};
+    if (chunk == 0) {
+      handler.SignalEnd();
+      break;
+    }
+    if (chunk < 0) {
+      auto err{errno};
+      if (err != EAGAIN && err != EWOULDBLOCK && err != EINTR) {
+        handler.SignalError(err);
+        break;
+      }
+    } else {
+      position_ += chunk;
+      got += chunk;
+    }
+  }
+  return got;
+}
+
+std::size_t OpenFile::Write(
+    Offset at, const char *buffer, std::size_t bytes, IoErrorHandler &handler) {
+  if (bytes == 0) {
+    return 0;
+  }
+  CriticalSection criticalSection{lock_};
+  CheckOpen(handler);
+  if (!Seek(at, handler)) {
+    return 0;
+  }
+  std::size_t put{0};
+  while (put < bytes) {
+    auto chunk{::write(fd_, buffer + put, bytes - put)};
+    if (chunk >= 0) {
+      position_ += chunk;
+      put += chunk;
+    } else {
+      auto err{errno};
+      if (err != EAGAIN && err != EWOULDBLOCK && err != EINTR) {
+        handler.SignalError(err);
+        break;
+      }
+    }
+  }
+  if (knownSize_ && position_ > *knownSize_) {
+    knownSize_ = position_;
+  }
+  return put;
+}
+
+void OpenFile::Truncate(Offset at, IoErrorHandler &handler) {
+  CriticalSection criticalSection{lock_};
+  CheckOpen(handler);
+  if (!knownSize_ || *knownSize_ != at) {
+    if (::ftruncate(fd_, at) != 0) {
+      handler.SignalErrno();
+    }
+    knownSize_ = at;
+  }
+}
+
+int OpenFile::ReadAsynchronously(
```

Comment:
Fortran standard says regarding asynchronous I/O in 12.6.4.1 par 4: 
```
For an asynchronous data transfer statement, the data transfers may occur during execution of the statement,
during execution of the corresponding wait operation, or anywhere between. The data transfer operation is
considered to be pending until a corresponding wait operation is performed.
```
It looks like to me that with `ReadAsynchronously` the data transfer occurs during execution of the statement. Whatever the implementation is doing, I think a comment to tell which option is implemented would be nice.